enhancements/operator-conditions-lib.md

```diff
@@ -0,0 +1,144 @@
+  namespace: operators
+status:
+  conditions:
+    type: Upgradeable # The name of the `foo` OLM Supported Condition.
+    status: "False"   # The operator is not ready to be upgraded.
+    reason: "migration"
+    message: "The operator is performing a migration."
+    lastTransitionTime: "2020-08-24T23:15:55Z"
+```
+
+The operator will be allowed to only modify the [`status`][status-sub-resource] sub-resource of the CR. Operators can either add, delete or update the `status.conditions` array to include OLM-supported or custom conditions. The format and description of the fields present in the conditions can be referred [here][condition_type].
+
+### Scenarios:
+
+1. Operator does not set/update any conditions that affect OLM:
+
+This is a no-op. In this case, OLM will default to its normal behavior.
+
+2. Operator tries to get/update the `Condition` CR when it is not present in the cluster:
+
+In this case, the specific API will return an error. The returned error can be wrapped, such that users can utilize the `errors.Is` method to compare the obtained error with specific values.
+
+3. Operator tries to get/update condition CRD which is present in cluster, but the requested CR instance is not available.
+
+In this case, the error returned will be in the same format as stated in the previous scenario but would specifically mention that the requested object is not found.
+
+To summarize, these are the overall helper methods/functions present in the library. For the ease of understanding, in the below table:
+- `Status` refers to the CR definition provided by OLM and 
+- [`metav1.Condition`][metav1-Condition] refers to specific Condition type present in `status.conditions` array.
+
+ Description             | Input(s)      | Return type(s) |  Expected output  | Remarks    |
+-------------            | ------------- | -------------| ----------------  |------------- |
+  GetStatusCRD        | Context, Runtime Object (`Status`), Object Key | `Status` struct, error | `Status` provided in input populated with values obtained from cluster |  1. Library will have helper functions to construct Object key <br/>2. Return errors as discussed in Scenarios section |
+  SetConditionStatus     |`metav1.Condition` to add or update | error | 1. Input condition added to `Status`<br/>2. Update `metav1.Condition.status` if already exists | Return errors as discussed in Scenarios section |
+  RemoveConditionStatus  | `metav1.Condition` to delete | error |Delete condition from `status.conditions` array | Return errors as discussed in Scenarios section |
+  FindConditionStatus   | `metav1.condition.Type` | `metav1.Condition` | Return `metav1.Condition` if exists, else `nil` | 
+  IsConditionStatusTrue | `metav1.condition.Type` | bool |`True` if `metav1.condition.Status` is set | 
+  IsConditionStatusPresentAndEqual | `metav1.condition.Type`, `metav1.condition.Status`| bool |`True` if `metav1.condition.Type` is present and status is equal as provided | 
```

Comment:
`FindConditionStatus()` is sufficient to check for condition truthiness/presence IMO
```suggestion
```

Reply:
Would these be good to have - if I'd like to check a particular `ConditionType`, instead of getting the entire condition? Included these two from: https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go#L84-L101

Reply:
Couldn't you just use the meta library functions directly if you pass `apiv1alpha1.Condition.Status.Conditions`?

Reply:
Yeah, but for the rest of the functions too, we are going to utilize the library functions. All of them would be wrappers around meta library functions.

Reply:
Yes, that could be done too. But then, all of the functions are going to be wrappers around meta library. In the sense instead of `FindConditionStatus()` we could use the one provided in meta library and pass `apiv1alpha1.Condition.Status.conditions`